### PR TITLE
listview fixes

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): closes unoplatform/kahua-private#113

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the new behavior?
1. Selector will attempt to bring selected item into the view, when it is set.
2. Added support for ScrollIntoView on skia and wasm platforms
3. Fixed the android impl of ScrollIntoView where it would leave the ScrollView in a corrupted state
4. Fixed a TabView bug where removing a tab ahead of the selected tab would corrupted the TabView content.
	^ ContainerFromItem returning the old index when called between collection mutation & next layout cycle.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TabView.cs [hunk 1]
```diff
@@ -44,10 +52,44 @@ public async Task When_First_Item_Unselected()
...
+
+		var container4 = setup.ContainerFromItem(4) as TabViewItem;
+		container4.DataContextChanged += (s, e) =>
+		{
+		};
+
+		// Remove a nearby materialized item that is before the selected item.
+		// On android, this will cause a ContainerFromItem bug to manifest
+		// where the items are no longer in sync with the materialized containers.
+		// This is guarantee to repro if both removed & selected items are currently materialized.
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TabView.cs:89]:
```suggestion
		// This is guaranteed to repro if both removed & selected items are currently materialized.
```

File: src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TabView.cs [hunk 2]
```diff
@@ -44,10 +52,44 @@ public async Task When_First_Item_Unselected()
...
+		};
+		setup.TabCloseRequested += (s, e) => source.Remove((int)e.Item);
+
+		setup.SelectedItem = 5;
+		await WindowHelper.WaitForIdle();
+
+		var container4 = setup.ContainerFromItem(4) as TabViewItem;
+		container4.DataContextChanged += (s, e) =>
+		{
+		};
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Microsoft_UI_Xaml_Controls/Given_TabView.cs:84]: debug code
```suggestion
		var presenter = setup.FindFirstDescendant<ContentPresenter>("TabContentPresenter");
		setup.TabCloseRequested += (s, e) => source.Remove((int)e.Item);

		setup.SelectedItem = 5;
		await WindowHelper.WaitForIdle();
```

File: src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs [hunk 3]
```diff
@@ -749,18 +747,21 @@ private double EstimatePanelExtent()
 
 		private void UpdateAverageLineHeight()
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.managed.cs:748]:
```
UpdateAverageLineHeight() => ...
CalculatePanelMeasureBreadth =>
	#if __WASM__ {...}
	...;
```

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs [hunk 4]
```diff
@@ -4428,6 +4428,62 @@ public async Task When_HeaderDataContext_Cleared_FromNavigation()
...
+			FrameworkElement CreateSetup() => typeof(T).Name switch
+			{
+				nameof(ListView) => new ListView { Width = 400, Height = 200, ItemsSource = source },
+				nameof(TabView) => new TabView { Width = 400, Height = 200, TabItemsSource = source },
+
+				_ => throw new ArgumentOutOfRangeException($"Generic arg not accepted: {typeof(T).Name}")
+			};
+			void SetSelectedItem(object item)
+			{
+				if (setup is ListView lv) lv.SelectedItem = item;
```
Comment (jeromelaban) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ListViewBase.cs:4482]: 🤖

**Review by jeromelaban — APPROVED** [review]
Nice!! Should we add a manual test for this scenario as well?

File: src/Uno.UI/Extensions/ViewExtensions.visual-tree.cs [hunk 5]
```diff
@@ -157,6 +197,17 @@ IEnumerable<string> GetDetails()
 		.OfType<T>()
 		.FirstOrDefault(predicate);
 
+	/// <summary>
+	/// Returns the first descendant of a specified type.
+	/// </summary>
+	/// <typeparam name="T">The type of descendant to find.</typeparam>
+	/// <param name="reference">Any node of the visual tree</param>
+	/// <param name="name">x:Name of the node</param>
+	/// <remarks>The crawling is done depth first.</remarks>
```
Comment (agneszitte) [inline src/Uno.UI/Extensions/ViewExtensions.visual-tree.cs:206]:
```suggestion
	/// <remarks>The crawling is done in depth first.</remarks>
```

File: src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.Android.cs [hunk 6]
```diff
@@ -302,58 +302,98 @@ private void ApplyScrollToPosition(int targetPosition, ScrollIntoViewAlignment a
...
 			if (alignment == ScrollIntoViewAlignment.Leading)
 			{
 				// 'Leading' means that the item always snaps to the top of the viewport no matter what
 				shouldSnapToStart = true;
 				shouldSnapToEnd = false;
 			}
+			else if (!shouldSnapToEnd && !shouldSnapToStart &&
+				FirstVisibleIndex == 0 && LastVisibleIndex + 1 == ItemCount)
+			{
+				// The item may be within materialized ranged (having not scroll at all in step-1) or
```
Comment (agneszitte) [inline src/Uno.UI/UI/Xaml/Controls/ListViewBase/VirtualizingPanelLayout.Android.cs:345]:
```suggestion
				// The item may be within materialized range (having not to scroll at all in step-1) or
```

**Review by agneszitte — APPROVED** [review]

**Comment by Xiaoy312** [comment]
last completed ci build (failing): https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=118526&view=results

**Pull request merged**